### ISSUE TITLE
Is it possible to change default directory (obj/docker/publish) to copy files?

### ISSUE THREAD

**Issue by yamac-kurtulus** [Microsoft/DockerTools#178]
I am working on a multi project solution, where all projects run inside a docker container using the docker-compose file. This worked fine, however I would like to separate my build context from the rest of my code, as it is too big. 

Therefore I have created a directory called `Docker` for each project and placed Dockerfiles in this directory. I want Visual Studio to output to this directory rather than the default `obj/docker/publish` directory when I press play on VS. Is it possible? I prefer to do it without modifying the project files too much.

**Comment by sgreenmsft** [comment]
Are all of the projects ASP.NET (.NET Framework)?

**Comment by yamac-kurtulus** [comment]
.Net-core

**Comment by sgreenmsft** [comment]
Thanks.  I was a bit uncertain regarding .NET Core or .NET Framework because the obj\docker\publish directory isn't used for .NET Core except for projects created by old versions of Visual Studio.  (I don't recall exactly but I think .NET Core projects stopped using obj\docker\publish around the time of VS2017 v15.4).

The general guidance for a large build context is to ensure you have a dockerignore file at the root of the build context and to update it to remove as much unnecessary content as possible.

Regarding moving the Dockerfile to another directory: docker-compose projects require that the Dockerfile be located in the same directory as the csproj (or, in the case of VB, the vbproj).  Note, though, that the Dockerfile's directory and the build context directory don't have to be identical.  In fact, for .NET Core, VS places the Dockerfile in the project directory and the build-context is the solution directory.

Is it possible to fix your build-context problem by updating the dockerignore file?

**Comment by yamac-kurtulus** [comment]
Thanks for explanation. My project still uses obj/docker/publish directory though. How do the new projects handle the source files then?

Also, I was able to change the context by modifying the docker-compose.yml file. E.g.
````
build:
    context: ./Admin/Admin/Docker
    dockerfile: Dockerfile
````
 It finds the context and the docker file in this case, but I need to move my assemblies and wwwroot to a directory in `Admin/Admin/Docker` So the context remains the same. Dockerfile is inside the Docker folder as well. 

I have a feeling that .dockerignore file is not used while using compose, because it takes too much time with or without ignore. Did not test it though.

**Comment by sgreenmsft** [comment]
Could you verify that the .dockerignore file has been placed in the root of the build context?

**Comment by yamac-kurtulus** [comment]
It is there indeed 
```
*
!obj/Docker/publish/*
!obj/Docker/empty/
```

**Comment by yamac-kurtulus** [comment]
Shouldn't I use publish by the way? The Publish directory and release directory have vast size difference.

**Comment by sgreenmsft** [comment]
Hmm.  Your scenario isn't quite clear to me.  Would you be able to share the full paths of these files in your solution?

sln, csproj, dcproj, docker-compose.yml, dockerfile, dockerignore

**Comment by yamac-kurtulus** [comment]
Right. Let me try it. 
```
RootFolder
+--solution.sln
+--docker-compose.yml
+--docker-compose.dcproj
+--ProjectA
      +--ProjectA
              +--ProjectA.dcproj
              +--Dockerfile
              +--.dockerignore
      +--ProjectA.Tests            //irrelevant
+--ProjectB
      +--ProjectB
              +--ProjectB.dcproj
              +--Dockerfile
              +--.dockerignore
      +--ProjectB.Tests            //irrelevant
+--ProjectC
...          //continues with other projects following the same structure.
```
Now here's the compose file:
```
version: '3.4'
services:
  projectA:
    image: repo/projectA
    build:
      context: ./projectA/projectA
      dockerfile: Dockerfile

  projectB:
    image: repo/projectB
    build:
      context: ./projectB/projectB
      dockerfile: Dockerfile
```
Here is Dockerfile for each project. Oh by the way I work on Windows 10 and our Servers are WS2016 without Fall creator's update. These were auto generated i believe. 
```
FROM microsoft/dotnet   //This repo is about to be obsolete i guess. 
                               //But I cannot use the current official one, because of our servers. 
ARG source
WORKDIR /app
EXPOSE 80
COPY ${source:-obj/Docker/publish} .
ENTRYPOINT ["dotnet", "`ProjectA.dll"]
```

On CI pipeline, i build each project manually with dotnet publish -o $projectWorkDir/obj/Docker/publish . I think best practice is to build inside container to speed up the build. Still the context seems too big. And copying a simple ~70MB directory for all projects takes 30 minutes.  

What I actually want to do is to be able to use this structure:
```
RootFolder
+--solution.sln
+--docker-compose.yml
+--docker-compose.dcproj
+--ProjectA
      +--ProjectA
              +--ProjectA.dcproj
              +--Docker
                  +--Dockerfile
                  +--.dockerignore
```

And publish to ProjectA/Docker directory

**Comment by sgreenmsft** [comment]
A couple of thoughts:
1) You'll not be able to move the dockerfile to another folder. It must remain beside the csproj for its associated project.
2) You can modify the build context to be any directory you'd like, however it's unclear to me why that would impact build performance, as the dockerignore file at the root of the build context excludes every file except the publish directory.
3) For CI/CD scenarios you should just be able to compile your sln file in release configuration.  The release build of the dcproj will build the images.

**Comment by yamac-kurtulus** [comment]
> 1. You'll not be able to move the dockerfile to another folder. It must remain beside the csproj for its associated project.

Well, this was specifically what I was asking. Thank you. 

> 2. You can modify the build context to be any directory you'd like, however it's unclear to me why that would impact build performance, as the dockerignore file at the root of the build context excludes every file except the publish directory.

Yeah just tested it, and apparently it works, but it still does not explain why my builds are sooo slow. Anyway. I'll play with the Dockerfile.

> 3. For CI/CD scenarios you should just be able to compile your sln file in release configuration.  The release build of the dcproj will build the images.

I am doing this step manually using docker-compose rather than using the project file. This is partly because the creation date of the project goes way back. Worth giving a shot though.

Thanks for the insights here. But I am still not sure about one thing. What is the best practice on building Docker images, if you know it. Should I prepare another dockerfile to build, so that I can abuse the cache in multistage builds like this:
```
FROM microsoft/aspnetcore:2.0 AS base
WORKDIR /app
EXPOSE 80

FROM microsoft/aspnetcore-build:2.0 AS builder
WORKDIR /src
COPY *.sln ./
COPY Web/Web.csproj Web/
RUN dotnet restore
COPY . .
WORKDIR /src/Web
RUN dotnet build -c Release -o /app

FROM builder AS publish
RUN dotnet publish -c Release -o /app

FROM base AS production
WORKDIR /app
COPY --from=publish /app .
ENTRYPOINT ["dotnet", "Web.dll"]
```
In this case I am afraid my context would require a bit micro management, and also does the debugger work while developing in this case, or would I need 2 Dockerfiles?

**Comment by sgreenmsft** [comment]
Yes, switching to a multistage dockerfile like that is a fine approach.  In general, though, I would not expect a multistage dockerfile like that to improve build performance -- actually I would expect it to be much slower than building on the host.

In your specific case, though, the slow docker build is a mystery so who knows whether the overall build would be faster.  =^)

Regarding the debugger -- for .NET Core projects, Visual Studio creates very similar multistage dockerfiles by default as of VS2017 v15.5 (or thereabouts).  The debugger works fine with them.

A special note about multistage dockerfiles: to make debugging super fast, Visual Studio does some magic with multistage dockerfiles.  When F5ing in *debug* configuration, VS builds only the first stage of the multistage dockerfile (i.e. the base stage), so none of the subsequent stages are run (i.e. the docker build doesn't run dotnet restore, dotnet build, or dotnet publish).  Instead, the essentially empty image created by the base stage is run, and VS mounts your project's output directory into the running container.  This allows VS to re-use a single running container for multiple F5s without having to perform a docker build operation for each code change you make.

tl;dr - Yes, debugging works well and we've made it super fast.  =^)

**Comment by yamac-kurtulus** [comment]
@sgreenmsft good to know and thanks for all the info. In my specific case, curiously, copying the files takes the most time. 100MB in 5 minutes... Therefore I thought of running a `dotnet restore` inside container would benefit from caching.

Closing this now. Thank you again.
